### DISCUSSION TITLE
Dynamic SQL

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#4560] [General]
Moving from #4539 as this may be useful for others...

To create a Server component for a module which uses dynamic SQL for data access:

*Note that you will need to include a package reference to Oqtane.Server in your modules's Client project (the reference does not exist because the default module template is designed to keep the Client and Server concerns separated from one another):

```
<PackageReference Include="Oqtane.Server" Version="5.2.0" />
```

Index.razor:

```
@namespace Siliqon.Module.Test521
@inherits ModuleBase
@using Oqtane.Repository
@inject ITenantRepository TenantRepository
@inject ISqlRepository SqlRepository

@if (tenant != null)
{
    <p>Tenant: @tenant.Name</p>
}

<button class="btn btn-primary" @onclick="Select">Select</button>

<button class="btn btn-primary" @onclick="Insert">Insert</button>

<p>@message</p>

@code {
    private Tenant tenant;
    private string message;

    protected override void OnInitialized()
    {
        tenant = TenantRepository.GetTenant(PageState.Site.TenantId);
    }

    private void Select()
    {
        var dr = SqlRepository.ExecuteReader(tenant, $"SELECT * FROM Test521");
        if (dr.Read())
        {
            message = dr["Name"];
        }
        else
        {
            message = "Test521 Does Not Exist";
        }
    }

    private void Insert()
    {
     SqlRepository.ExecuteNonQuery(tenant, $"INSERT INTO Test521(Name, CreatedOn, CreatedBy, ModifiedOn, ModifiedBy) VALUES('Test', getdate(),   '{PageState.User.Username}', getdate(), '{PageState.User.Username}')");
    }
}
```

Note that your razor component is running on the server in this example and therefore is able to access the server Repository methods directly. The module project can be trimmed down to the following files in this case:

![image](https://github.com/user-attachments/assets/d49f982d-1e95-4d42-a3c0-56635a892646)

The huge caveat to using this approach is that the module will not run on WebAssembly or .NET MAUI - it will only work in a Server run-time scenario. It will also only run on SQL Server.

**Comment by ADefWebserver** [comment]
The [DAL+](https://www.dnnsoftware.com/community-blog/cid/136768/dal-dal-connect-your-dotnetnuke-module-to-the-database) returns... 😉

**Reply by sbwalker** [reply]
LOL! This capability has existed since Oqtane 1.0 however there were no examples because it's not really a recommended practice anymore.

**Reply by hishamco** [reply]
It's useful in some cases, I need to have a look into the implementation
